torch/ao/pruning/_experimental/pruner/README.md

```diff
@@ -1,93 +1,263 @@
 
-* when provided in `config` with corresponding Conv2d layer
+```python
+pruner.prepare(model, sparse_config)
+```
 
-`BasePruner`:  base class with abstract method `update_mask` that computes the new pruner mask for all modules (see Write Your Own Pruner). The base pruner prunes the entire channel by default (weight & corresponding bias); if you don’t want the bias to be pruned, then set `also_prune_bias` to be False.
+However, if we were to run the prepared model currently, we'd see the exact same output as before the prepared model.
 
-`prepare`: registers the pruning parametrization (called `PruningParametrization`) to each module layer of the model; also adds forward hooks for bias support and re-inserting zeros to the output so the next layer received the correct size input.
+This is because we have yet to update the mask, and by default the initial mask is all set to `True` and prunes no rows.
+We can calculate the mask given the current weights by takeing a pruning step.
 
-Note: for BatchNorm2d layers, the parametrization `ZeroesParametrization` is attached instead since its weight is 1d, so removing channels would affect the input dimension as well. `ZeroesParametrization` zeroes out channels rather than removing them like `PruningParametrization`. We need this when `also_prune_bias=True`, so BatchNorm2d channels get pruned with their corresponding Conv2d channels.
+```python
+# need to enable mask update
+pruner.enable_mask_update = True
+pruner.step()
+```
 
+After this step is called, the masks will be updated to prune the corresponding weights. You can expect the ouput of the model after this step to be the same as the pruned model.
 
+However, there will be no performance / memory benefits for the model, since the pruned parameters are still present but just zeroed out. In order to achieve performance gains, we need to resize the weight tensors, by calling `prune`.
+```python
+# to prune the module
+pruned_model = pruner.prune()
 ```
-pruner = ImplementedPruner(defaults=None, also_prune_bias=True)
-pruner.prepare(model, config)
-```
 
+Prune will symbolically trace the model in order to get a graph, and then search that graph for particular patterns (op sequences)
+
+Once prune identifies an op sequence, it is able to resize the weights by calling the pruning function for a paticular block. This is defined by a dictionary that maps patterns (tuple of nn modules / functions) to a particular pruning function. You can extend this pattern dictionary to add support for pruning different modules / patterns.
 
-`step`: applies `update_mask` logic (i.e. prunes the weight matrix)
+For this particular example, we would expect to match the `(nn.Linear, nn.ReLU, nn.Linear)` pattern in our default pattern dict.
 
+Once this is done for all matches, `prune` returns your pruned model:
```

Comment:
I would maybe clarify that this is not actually what is being returned, it returns a pruned model, the following are useful stats to compare the before/after